controllers/servingruntime_validator.go

```diff
@@ -0,0 +1,154 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package controllers
+
+import (
+	"fmt"
+	"strings"
+
+	"errors"
+
+	corev1 "k8s.io/api/core/v1"
+
+	api "github.com/kserve/modelmesh-serving/apis/serving/v1alpha1"
+	"github.com/kserve/modelmesh-serving/controllers/modelmesh"
+)
+
+// validateServingRuntimeSpec returns an error if the spec is invalid
+// Checks include:
+// - BuiltInAdapter has a container matching the name of the type
+// - names of entries do not overlap with reserved names
+// - spec does not override required internals like Volumes or Containers
+// - containers do not mount internal only Volumes
+// - some fields in containers are controlled by model mesh and cannot be set
+// - check for overlaps in declared ports with internal ports
+func validateServingRuntimeSpec(rt *api.ServingRuntime) error {
+	return validationChain(rt,
+		validateBuiltInAdapterSpec,
+		validateContainers,
+		validateVolumes,
+	)
+}
+
+func validationChain(rt *api.ServingRuntime, funcs ...func(*api.ServingRuntime) error) error {
+	for _, f := range funcs {
+		if err := f(rt); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func validateBuiltInAdapterSpec(rt *api.ServingRuntime) error {
+	if rt.Spec.BuiltInAdapter == nil {
+		return nil // nothing to check
+	}
+
+	st := rt.Spec.BuiltInAdapter.ServerType
+	if _, ok := builtInServerTypes[st]; !ok {
+		return fmt.Errorf("unrecognized built-in runtime server type %s", st)
+	}
+	for ic := range rt.Spec.Containers {
+		if rt.Spec.Containers[ic].Name == string(st) {
+			return nil // found, all good
+		}
+	}
+
+	return fmt.Errorf("must include runtime container with name %s", st)
+}
+
+func validateContainers(rt *api.ServingRuntime) error {
+	for i := range rt.Spec.Containers {
+		c := &rt.Spec.Containers[i]
+		if err := validateContainer(c); err != nil {
+			return fmt.Errorf("container spec for %s is invalid: %w", c.Name, err)
+		}
+	}
+	return nil
+}
+
+func validateContainer(c *corev1.Container) error {
+	// Block container names that conflict with injected containers or reserved prefixes
+	internalContainerNames := []string{
+		modelmesh.ModelMeshContainer,
+		modelmesh.RESTProxyContainer,
+		modelmesh.PullerContainer,
+	}
+	if err := checkName(c.Name, internalContainerNames, "container name"); err != nil {
+		return err
+	}
+
+	// TODO: block if container name matches `*-adapter`??
+
+	// Block fields required for model-mesh to control the pod lifecycle
+	if c.ReadinessProbe != nil || c.Lifecycle != nil {
+		return errors.New("ReadinessProbe and Lifecycle are managed by modelmesh and cannot be set")
+	}
```

Comment:
I'm not sure whether we should block these, at least not the liveness one. Though the desire is for model-mesh to own these and determine from the container itself, it might not do this well enough yet. I think some folks want to set their own.

Reply:
The Liveness probe is not restricted here as that can be set in the current pre-Container SR spec. My goal here was to match the current behavior but maintain current restrictions where they make sense.

If you want the injected values for ReadinessProbe and Lifecycle to be overridable, I will remove this.

Reply:
OK yeah I mistakenly read this as checking both probes. I guess this is ok, we can always open up further if needed in future.